express_relay/programs/express_relay/src/lib.rs

```diff
@@ -0,0 +1,366 @@
+        express_relay_metadata_data.fee_receiver_relayer = *ctx.accounts.fee_receiver_relayer.key;
+
+        Ok(())
+    }
+
+    // Sets a new set of fee splits for the program
+    pub fn set_splits(ctx: Context<SetSplits>, data: SetSplitsArgs) -> Result<()> {
+        validate_fee_split(data.split_protocol_default)?;
+        validate_fee_split(data.split_relayer)?;
+
+        let express_relay_metadata_data = &mut ctx.accounts.express_relay_metadata;
+
+        express_relay_metadata_data.split_protocol_default = data.split_protocol_default;
+        express_relay_metadata_data.split_relayer = data.split_relayer;
+
+        Ok(())
+    }
+
+    // Sets a fee split for a specific protocol
+    pub fn set_protocol_split(ctx: Context<SetProtocolSplit>, data: SetProtocolSplitArgs) -> Result<()> {
+        validate_fee_split(data.split_protocol)?;
+
+        ctx.accounts.protocol_config.protocol = *ctx.accounts.protocol.key;
+        ctx.accounts.protocol_config.split = data.split_protocol;
+
+        Ok(())
+    }
+
+    // Permissions a particular (protocol, permission) pair and distributes bids according to splits
+    pub fn permission(ctx: Context<Permission>, data: PermissionArgs) -> Result<()> {
+        if data.deadline < Clock::get()?.unix_timestamp {
+            return err!(ErrorCode::DeadlinePassed);
+        }
+
+        // check that not cpi
+        if get_stack_height() > TRANSACTION_LEVEL_STACK_HEIGHT {
+            return err!(ErrorCode::InvalidCPIPermission);
+        }
+
+        // handle bid payment
+        let bid_amount = data.bid_amount;
+        let searcher = &ctx.accounts.searcher;
+        let rent_searcher = Rent::get()?.minimum_balance(searcher.to_account_info().data_len());
+        if bid_amount + rent_searcher > searcher.lamports() {
+            return err!(ErrorCode::InsufficientSearcherFunds);
+        }
+
+        let express_relay_metadata = &ctx.accounts.express_relay_metadata;
+        let split_relayer = express_relay_metadata.split_relayer;
+        let split_protocol_default = express_relay_metadata.split_protocol_default;
+
+        let split_protocol: u64;
+        let protocol_config = &ctx.accounts.protocol_config;
+        let protocol_config_account_info = protocol_config.to_account_info();
+        if protocol_config_account_info.data_len() > 0 {
+            let account_data = &mut &**protocol_config_account_info.try_borrow_data()?;
+            let protocol_config_data = ConfigProtocol::try_deserialize(account_data)?;
+            split_protocol = protocol_config_data.split;
+        } else {
+            split_protocol = split_protocol_default;
+        }
+
+        let fee_protocol = bid_amount * split_protocol / FEE_SPLIT_PRECISION;
+        if fee_protocol > bid_amount {
+            return err!(ErrorCode::FeesHigherThanBid);
+        }
+
+        let fee_relayer = bid_amount.saturating_sub(fee_protocol) * split_relayer / FEE_SPLIT_PRECISION;
```

Comment:
the subtraction here seems like it will make it harder to configure the fees (?) I get that this makes it a cascade, but seems simpler to make the splits add up to 100% conceptually.

Reply:
that's fair. in a way i think it makes sense for the relayer split to be a percentage of what isn't taken by the protocol (so that relayer revenue remains linear with Express Relay revenue).

i think there are merits to both, but mostly, this structure is preferred bc it's consistent with the evm side